moto/core/models.py

```diff
@@ -427,7 +430,9 @@ class base_decorator:
     def __init__(self, backends: BackendDict):
         self.backends = backends
 
-    def __call__(self, func: Optional[Callable[..., Any]] = None) -> BaseMockAWS:
+    def __call__(
+        self, func: Optional[Callable[..., Any]] = None
+    ) -> Union[BaseMockAWS, Callable[..., BaseMockAWS]]:
```

Comment:
Same: if `func` is provided, we return a callable returning whatever type `func` returned, not `BaseMockAWS`.